### DISCUSSION TITLE
Running `quarto preview` should show localhost url more frequently

### DISCUSSION THREAD

**Discussion by kdheepak** [quarto-dev/quarto-cli#4779] [Feature Requests]
### Bug description

Currently, when running `quarto preview` for a while and editing multiple pages, there's lots of output on `stdout`. If I accidentally close the tab with the preview (and if I'm not able to reopen it because it's been a while), I'm not able to go back to the terminal and see the URL where the preview server is hosting the code. I have to either try to scroll all the way up (sometimes this has exceeded my scrollback buffer) or I have to restart `quarto preview` (which before using `freeze` and `cache` took a long time to execute).

One of these options would solve it for me - Either show the URL after updates, or allow the user to press a key that prints out the URL.

### Checklist

- [ ] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [ ] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [ ] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [ ] Please document the operating system you're running. If on Linux, please provide the specific distribution.

**Comment by mcanouil** [comment]
Which IDE are you using?
I am guessing VSCode, if that is the case you have a panel where you can find open and forwarded ports.

**Reply by kdheepak** [reply]
I use both VSCode and the terminal + neovim. I can probably find open and forwarded ports easily (on my mac at least, on windows I'll have to learn how to do this) but I think this would also be useful for less tech savvy folk.

**Reply by mcanouil** [reply]
I realised/noticed, the ports are not always showed in VSCode "ports" panel when working locally (using remote host SSH/container/WSL always display them).

I wonder, how do you run `quarto preview`? Depending on how you run it, you should not be able to use the same terminal to do something else, thus leaving the preview URL visible in the terminal.

**Reply by kdheepak** [reply]
The preview URL just gets lost to scrolling when I’m working while having it open for a long time.

**Reply by mcanouil** [reply]
I've got that, but I suspect you are typing the command yourself in the terminal, which leads to that behaviour.

For example, I tested on MacOS and Ubuntu, when the render button is used to trigger the rendering it leads to a "busy" terminal, thus you cannot loose the URI when scrolling (you actually cannot scroll down)
<img width="1624" alt="image" src="https://user-images.githubusercontent.com/8896044/225127285-392beee3-d83c-4b66-9773-e72a8af132b6.png">

EDIT: I have the same behaviour when typing `quarto preview demo.qmd` for both MacOS and Ubuntu.

**Reply by kdheepak** [reply]
I see. I wasn't used the render button in VSCode because I was getting errors. 

But even if I start with a single file in the terminal, if I'm editing multiple files, then the url is printed every time I visit a page:

<img width="750" alt="image" src="https://user-images.githubusercontent.com/1813121/225173618-650410d6-ae2a-4648-9c09-c851ccb7a78b.png">

It also appears to print the config it is using. When I revisit a page, it reprints it. Is this not the behavior for others?

```
$ quarto preview understanding-variables-and-mutability-in-julia/index.qmd
pandoc
  to: html
  from: markdown+emoji+mark
  output-file: index.html
  standalone: true
  title-prefix: kdheepak
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  css:
    - ../_quarto/styles.css
  email-obfuscation: javascript
  html-q-tags: true
  toc: true
  toc-depth: 3
  shift-heading-level-by: 1
  number-sections: false

metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  author: Dheepak Krishnamurthy
  quarto-required: '>= 1.3.0'
  links-as-notes: true
  crossref:
    fig-prefix: Figure
    tbl-prefix: Table
    eq-prefix: Equation
  jupyter: julia-1.8
  bibliography:
    - ../_quarto/blog.bib
  csl: ../_quarto/ieee.csl
  license: CC BY-SA
  grid:
    content-mode: full
  theme:
    light:
      - pulse
      - ../_quarto/theme.scss
    dark:
      - solar
      - ../_quarto/theme.scss
  code-copy: hover
  anchor-sections: true
  toc-location: left
  callout-appearance: simple
  title-block-banner: true
  title-block-categories: false
  citations-hover: true
  cap-location: margin
  citation: true
  comments:
    hypothesis: true
    giscus:
      repo: kdheepak/blog
      category: Comments
      reactions-enabled: true
      loading: lazy
      input-position: top
      mapping: url
  title: Understanding Variables and Mutability in Julia
  summary: 'In this blog post, I present a mental model for understanding the basics of variable assignment and mutability in Julia, using analogies to books and bookshelves to help illustrate the concepts.'
  date: '2023-02-25T21:14:15-0500'
  categories:
    - julia
  keywords: 'julia, analogy, pass by sharing'
  references:
    - id: ndarray
      title: Multi-dimensional Arrays
      URL: 'https://docs.julialang.org/en/v1/manual/arrays/'
    - id: functions
      title: Functions
      URL: 'https://docs.julialang.org/en/v1/manual/functions/'
  nocite: |
    @ndarray
    @functions


Output created: ../_site/understanding-variables-and-mutability-in-julia/index.html

Preparing to preview

Watching files for changes
Browse at http://localhost:3215/understanding-variables-and-mutability-in-julia/index.html

GET: /understanding-variables-and-mutability-in-julia/index.html

GET: /

GET: /installing-homebrew-in-an-alternative-location/
GET: /
GET: /understanding-variables-and-mutability-in-julia/

Starting julia-1.8 kernel...(Notebook read from cache)

pandoc
  to: html
  from: markdown+emoji+mark
  output-file: index.html
  standalone: true
  title-prefix: kdheepak
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  css:
    - ../_quarto/styles.css
  email-obfuscation: javascript
  html-q-tags: true
  toc: true
  toc-depth: 3
  shift-heading-level-by: 1
  number-sections: false

metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  author: Dheepak Krishnamurthy
  quarto-required: '>= 1.3.0'
  links-as-notes: true
  crossref:
    fig-prefix: Figure
    tbl-prefix: Table
    eq-prefix: Equation
  jupyter: julia-1.8
  bibliography:
    - ../_quarto/blog.bib
  csl: ../_quarto/ieee.csl
  license: CC BY-SA
  grid:
    content-mode: full
  theme:
    light:
      - pulse
      - ../_quarto/theme.scss
    dark:
      - solar
      - ../_quarto/theme.scss
  code-copy: hover
  anchor-sections: true
  toc-location: left
  callout-appearance: simple
  title-block-banner: true
  title-block-categories: false
  citations-hover: true
  cap-location: margin
  citation: true
  comments:
    hypothesis: true
    giscus:
      repo: kdheepak/blog
      category: Comments
      reactions-enabled: true
      loading: lazy
      input-position: top
      mapping: url
  title: Programmers' preferences for package names
  categories:
    - analysis
    - python
    - julia
    - rust
  keywords: 'python, julia, rust, R'
  summary: I was curious how package names were chosen in various language ecosystems.
  date: '2022-07-29T01:47:12-0600'


GET: /programmers-preferences-for-package-names/
```

Annoyingly I get an error when I use the render button inside VSCode:

```
quarto preview /Users/kd/gitrepos/blog/programmers-preferences-for-package-names/index.qmd --no-browser --no-watch-inputs
ERROR: Jupyter kernel 'julia-1.8' not found.

Error: Jupyter kernel 'julia-1.8' not found.
    at jupyterKernelspecFromMarkdown (file:///Users/USERNAME/Applications/quarto/bin/quarto.js:53319:35)
    at async quartoMdToJupyter (file:///Users/USERNAME/Applications/quarto/bin/quarto.js:53152:36)
    at async createNotebookforTarget (file:///Users/USERNAME/Applications/quarto/bin/quarto.js:55555:16)
    at async Object.target (file:///Users/USERNAME/Applications/quarto/bin/quarto.js:55348:18)
    at async fileExecutionEngineAndTarget (file:///Users/USERNAME/Applications/quarto/bin/quarto.js:58149:20)
    at async renderContexts (file:///Users/USERNAME/Applications/quarto/bin/quarto.js:93147:34)
    at async renderFormats (file:///Users/USERNAME/Applications/quarto/bin/quarto.js:95567:26)
    at async previewFormat (file:///Users/USERNAME/Applications/quarto/bin/quarto.js:106697:36)
    at async Command.fn (file:///Users/USERNAME/Applications/quarto/bin/quarto.js:107998:28)
    at async Command.execute (file:///Users/USERNAME/Applications/quarto/bin/quarto.js:8437:13)
```

It seems like my VSCode environment is different from my terminal environment, and homebrew gets picked up before miniconda3 in VSCode for some reason.

**Reply by mcanouil** [reply]
Oh I see. You are using `quarto preview` and watching input changes, but you close the preview for some reason.

In VSCode you can use the preview panel, you can even easily decrease panel size quite easily since February update.
You can also use your favourite browser to preview your Quarto documents.

What I don't understand is why are you using `quarto preview` if you are going to close the preview, thus the need to get the URI back at some point? It's quite an unintuitive use case to me.

You can change the on save render from the YAML header or Quarto extension settings.
Same as the "reveal preview panel on render" feature.
<img width="1161" alt="image" src="https://user-images.githubusercontent.com/8896044/225426775-05f9eedd-e41a-4c87-a2af-f355b6eae38c.png">

Small recording showing the preview panel and the VSCode panel focus/resize feature.

https://user-images.githubusercontent.com/8896044/225427124-7d2d511f-ec0b-4620-b2b4-6fef17a47b11.mov

**Reply by mcanouil** [reply]
If the render button does not have the same behaviour, this means your setup is off somehow which can lead to unexpected behaviour even using Quarto CLI from the terminal directly.

**Comment by kdheepak** [comment]
> What I don't understand is why are you using quarto preview if you are going to close the preview, thus the need to get the URI back at some point? It's quite an unintuitive use case to me.

I just close it accidentally. For context I was porting my blog to quarto over the weekend and writing a tutorial for Julia at work before that, in both cases I had `quarto preview` running for multiple days as I was writing and testing various features while also working on other things. And multiple times I accidentally closed the preview window in the browser. Maybe this is not a common usecase in regular use? I think when people first start using quarto porting existing material or creating lots of new material, I can see this being something that they come across.

**Reply by mcanouil** [reply]
Why not just stop preview and start a fresh one when you come back to work on it?
It's definitely better in my opinion to avoid weird behaviour which is solved by "turning it off and on again" 😅

IT Crowds: <https://youtu.be/nn2FB1P_Mn8>

**Reply by kdheepak** [reply]
The main reason was that when using `freeze: false` it would take a long time.

**Reply by mcanouil** [reply]
Maybe "spamming" the URI in stdout is not the solution.
Maybe some kind of "attach" behaviour to tie the preview URI to Quarto VSCode panel would be better to retrieve the URI.

For instance, with the R extension for VSCode you can open an R session from the terminal and manually attached it to the R extension which allow the extension to scan environment, etc.

**Reply by kdheepak** [reply]
Or if there’s a way to make quarto only print out the url instead of the entire config, that would also solve my problem. I can also just try to change my behavior. After I started using freeze I didn’t think this was a pain point at all.

**Reply by mcanouil** [reply]
Indeed, displaying metadata is probably too much in preview, especially if there are no changes in it.

**Comment by jimjam-slam** [comment]
FWIW I also tend to have this problem—I'm previewing website updates and making liberal use of hot reload, and then at some point I realise that I accidentally closed the tab a while ago.

The thing that is really helping me with this is working in development containers, which is a lot easier for me now that Quarto builds on aarch64 (and hence I can run R + Quarto devcontainers on Apple Silicon). When you're running in a dev container or on a remote host, VSCode catches the ports used by Quarto (or other web servers) and automatically forwards them to the host so that you can connect to them invisibly.

It _also_ has a Ports pane that maintains a list of forwarded ports, which processes opened them and which ones _continue_ to have a running processes attached, which makes it easy to get back to my preview if I lose it. There are even buttons to open  that port on the host in an external browser or in the VSCode preview if I prefer the latter and have closed it!

<img width="1611" alt="Screenshot of VSCode with a dev container open. The terminal shows the quarto preview command running with a web server running, and the Ports pane shows that it is forwarded to the host. A previously opened port with no running process is also shown, and distinguished from the current port by an open bullet. Buttons for opening the port in a browser or preview are visible on hover." src="https://user-images.githubusercontent.com/6520659/225486610-2a6f6912-265e-4ec2-a26f-47197bad906b.png">

Unfortunately you don't get this pane when developing locally, but it's very nice, and dev containers are much easier to use now!

**Reply by mcanouil** [reply]
Indeed, this is what I was talking about in my first reply (<https://github.com/quarto-dev/quarto-cli/discussions/4779#discussioncomment-5282024>).
Unfortunately, for now and as you mentioned, the panel is only for "remote" hosts.

**Reply by jimjam-slam** [reply]
Sorry @mcanouil, I missed your comment 😅

**Comment by gshotwell** [comment]
I run into this all the time, and I think it's worth considering for a few reasons: 
- When I'm working on a website I often want to launch it in a browser instead of rendering in the window, but I often accidentally lose the tab which has the preview running
- Not everyone uses VS Code, the docs point people to using `quarto preview` for rendering those websites
- I sometimes have a make process which ends with `quarto preview` which is hard to replicate with the render behaviour
- Sometimes I want to use the quarto preview CLI options

**Comment by mcanouil** [comment]
Quick note, the port can be set which will avoid having to remember the preview URL, see <https://quarto.org/docs/reference/projects/options.html#preview>.

Closing as:

- https://github.com/quarto-dev/quarto-cli/discussions/6890#discussioncomment-7037234

**Reply by kdheepak** [reply]
I also have multiple quarto websites I’m working on at the same time, also particularly when I’m collaborating with others. It’s not practical to change the port in some projects for me.

**Reply by mcanouil** [reply]
Not my call unfortunately.

**Comment by fisher-j** [comment]
I would also like to see the url reported more frequently. My current solution is to look through my browse history if the tab gets closed.
